crates/shielder-scheduler-server/src/storage/dynamo_db.rs

```diff
@@ -0,0 +1,475 @@
+use aws_sdk_dynamodb::{
+    error::SdkError,
+    operation::{describe_table::DescribeTableError, put_item::PutItemError},
+    types::{
+        AttributeDefinition, AttributeValue, GlobalSecondaryIndex, KeySchemaElement, KeyType,
+        Projection, ProjectionType, ProvisionedThroughput, ScalarAttributeType,
+    },
+    Client,
+};
+use chrono::{DateTime, Utc};
+
+use crate::storage::{RequestStatus, ScheduledRequest, StorageError, StorageInterface};
+
+const TABLE_NAME: &str = "ScheduledRequests";
+const STATUS_RELAY_INDEX: &str = "StatusRelayAfterIndex";
+
+pub struct DynamoDb {
+    client: Client,
+}
+
+impl DynamoDb {
+    pub async fn new() -> Result<Self, StorageError> {
+        let config = aws_config::load_defaults(aws_config::BehaviorVersion::latest()).await;
+        let client = Client::new(&config);
+        let db = Self { client };
+        db.create_table_if_not_exists().await?;
+        Ok(db)
+    }
+
+    async fn create_table_if_not_exists(&self) -> Result<(), StorageError> {
+        // Describe the table. If it exists return Ok
+        match self
+            .client
+            .describe_table()
+            .table_name(TABLE_NAME)
+            .send()
+            .await
+        {
+            Ok(_) => return Ok(()),
+            Err(e) => {
+                if !is_resource_not_found(&e) {
+                    return Err(map_internal("describe_table", e));
+                }
+            }
+        }
+
+        let attr_def_id = AttributeDefinition::builder()
+            .attribute_name("id")
+            .attribute_type(ScalarAttributeType::S)
+            .build()
+            .map_err(|e| StorageError::Internal(format!("AttrDef build error: {e}")))?;
+
+        let attr_def_status = AttributeDefinition::builder()
+            .attribute_name("status")
+            .attribute_type(ScalarAttributeType::S)
+            .build()
+            .map_err(|e| StorageError::Internal(format!("AttrDef build error: {e}")))?;
+
+        let attr_def_relay_after = AttributeDefinition::builder()
+            .attribute_name("relay_after")
+            .attribute_type(ScalarAttributeType::N)
+            .build()
+            .map_err(|e| StorageError::Internal(format!("AttrDef build error: {e}")))?;
+
+        let key_schema = KeySchemaElement::builder()
```

Comment:
This is good since
> If your table has a simple primary key (partition key only), DynamoDB stores and retrieves each item based on its partition key value.

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.Partitions.html